File: list/src/main/java/ArrayList.java
public class ArrayList implements List {
    //线性表的大小
    protected int size;
    //线性表的最大长度
    protected int maxLen;
    //线性表的元素集合
    protected Object[] items;
    //线性表的默认长度
    protected static final int DEFAULT_LENGTH = 3;

    ArrayList(){
        this.size = 0;
        this.items = new Object[DEFAULT_LENGTH];
        this.maxLen = DEFAULT_LENGTH;
    }
    public void clear() {
        this.size = 0;
    }

    public boolean isEmpty() {
        return this.size==0?true:false;
    }

    public int length() {
        return size;
    }

    public Object get(int index) {
        if(index <0 || index > size-1){
            throw new IndexOutOfBoundsException("illegal index:" +index+">"+size);
        }
        return items[index];
    }

    public int indexOf(Object o) {
        if(o == null){
            throw new IllegalArgumentException("illegal item:"+o);
        }
        int i = 0;
        for (Object item:items){
            if(o.equals(item)){
                return i;
            }
            i++;
        }
        return -1;
    }

    public Object pre(Object o) {
        int index = indexOf(o);
        if(index == -1 || index == 0){
            return null;
        }else{
            return items[index - 1];
        }
    }

    public Object next(Object o) {
        int index = indexOf(o);
        if(index == -1 || index == size -1){
            return null;
        }else{
            return items[index + 1];
        }
    }

    public void insert(int index, Object o) {
        if(index< 0 || index > size){
            throw new IndexOutOfBoundsException("illegal index:"+index);
        }
        if(size >= maxLen-1){
            expand();
        }
        for(int i=size; i>= 0; i--){
            if(i==index){
                items[i] = o;
                break;
            }else{
                items[i]=items[i-1];
            }
        }
        size++;
    }

    public void delete(int index) {
        if(index < 0 || index > size-1){
            throw new IndexOutOfBoundsException("illegal index:"+ index);
        }
        for(int i = index; i < size-1;i++){
            items[i]=items[i+1];
        }
        size--;
    }

    public void display() {
        int i = 0;
        for (Object item:
             items) {
            if(i == size){
                System.out.println();
                break;
            }
            System.out.print(item + " ");
            i++;
        }
    }

    protected void expand(){
        maxLen = 2*maxLen;
        Object[] newItems = new Object[maxLen];
        int i = 0;
        for (Object item:items) {
            newItems[i] = item;
            i++;
        }
        items = newItems;
    }

    public void union(List lb) {
        for(int i = 0; i<lb.length();i++){
            if(this.indexOf(lb.get(i)) == -1){
                this.insert(size,lb.get(i));
            }
        }
    }

    public ArrayList merge(List lb) {
        ArrayList newList = new ArrayList();
        int length = lb.length();
        int i,j;
        for( i = size-1,j = length-1;i>=0 && j>=0;){
            if((Integer)items[i]<(Integer) lb.get(j)){
                newList.insert(0,lb.get(j));
                j--;
            }else{
                newList.insert(0,items[i]);
                i--;
            }
        }
        if (i < 0 && j >= 0){
            for(int k=j;k>=0;k--){
                newList.insert(0,lb.get(k));
            }
        }
        if(i >= 0 ){
            for(int k=i;k>=0;k--){
                newList.insert(0,items[k]);
            }
        }
        return newList;
    }
}

File: Python/list/ArrayList.py
# -*- utf-8 -*-


class ArrayList:

    def __init__(self):
        self.__capacity = 10
        self.__size = 0
        self.__items = [0]*self.__capacity

    def index_of(self, item):
        for i in range(self.__size):
            if item == self.__items[i]:
                return i
        return -1

    def add(self, item):
        if self.__size == self.__capacity:
            self.__expand()
        self.__items[self.__size] = item
        self.__size += 1

    def insert(self, index, item):
        if index > self.__size:
            return;
        if self.__size == self.__capacity:
            self.__expand()
        self.__items[index+1:self.__size+1] = self.__items[index:self.__size]
        self.__items[index] = item
        self.__size += 1

    def delete(self, item):
        for i in range(self.__size):
            if item == self.__items[i]:
                self.__items[i:self.__size-1] = self.__items[i+1:self.__size]
                self.__size -= 1

    def size(self):
        return self.__size

    def is_empty(self):
        return True if self.__size == 0 else False

    def display(self):
        for i in range(self.__size):
            print(self.__items[i], end=" ")
        print()

    def __expand(self):
        self.__capacity *= 2
        temp = [0]*self.__capacity
        temp[0:self.__size] = self.__items[:]
        self.__items = temp


if __name__ == '__main__':
    al = ArrayList()
    print(al.size())
    print(al.is_empty())
    al.add(1)
    al.insert(0, 2)
    al.insert(2, 3)
    al.display()
    print(al.size())
    al.delete(1)
    al.display()
    for i in range(10):
        al.insert(0, i)
    al.display()
